### DISCUSSION TITLE
Configuration for an entire python package

### DISCUSSION THREAD

**Discussion by jelkuweiss** [facebookresearch/hydra#2310] [Q&A]
Hello!

I wanted to know if there is a standard way(or an example) of using this hydra config for a python package that is then imported into the main script. 

If I have a package made of a number of files and subfiles of scripts containing functions, is there a way to edit the inputs or parameters for these function directly in the package rather than passing it on from the main execution?

**Comment by Jasha10** [comment]
Hi @jelkuweiss,
The two standard entrypoints for Hydra are `hydra.main` and `hydra.utils.compose`.
You can decorate as many functions with `@hydra.main` as you like, and you can use `compose` as many times as you like.

Does this answer your question?

**Reply by jelkuweiss** [reply]
Hey, Thanks for getting back to me.
So I tried the hydra.compose and it works well, but if I have a plugin in the config (Using the Optuna one) it is not working well.

Say I have a package and I pass all the functions in it the configuration using 'compose', and then I call that function (consider that all it returns is just the input of one float in the configuration, that we are trying to minimise using optuna, so it should just pick the lowest value of it) in another script where my main is and return exactly that function. What I am getting from the plugin is the value of the default config rather than the optimal one.

**Reply by Jasha10** [reply]
The compose API does not implement support for multirun; only a single config object is "composed". Meanwhile, the optuna sweeper plugin relies on multiple runs (i.e. multiple calls to your main function); the optuna plugin only works with `@hydra.main`.

> in another script where my main is and return exactly that function

The compose API is a functional API (in the sense of functional programming). It does not call back into user code (i.e. it will not call your main function defined in another script). All that `compose` does is invoke Hydra's machinery to put together yaml files.

**Reply by Jasha10** [reply]
Check out [`hydra_zen.launch`](https://mit-ll-responsible-ai.github.io/hydra-zen/generated/hydra_zen.launch.html), which implements a functional API on top of `hydra.main`. It might be closer to what you are looking for; it has support for multirun and plugins.
